cpp/src/utils/metrics/yaml_config.cc

```diff
@@ -0,0 +1,192 @@
+  }
+
+  if (system_metrics_node["gauge"]) {
+    const std::vector<MetricConfiguration> gauge_metrics =
+        ParseMetrics(system_metrics_node["gauge"], MetricType::GAUGE);
+    system_metrics.insert(system_metrics.end(), gauge_metrics.begin(),
+                          gauge_metrics.end());
+  }
+
+  if (system_metrics_node["histogram"]) {
+    const std::vector<MetricConfiguration> histogram_metrics =
+        ParseMetrics(system_metrics_node["histogram"], MetricType::HISTOGRAM);
+    system_metrics.insert(system_metrics.end(), histogram_metrics.begin(),
+                          histogram_metrics.end());
+  }
+}
+
+void YAMLMetricsConfigurationHandler::ParseModelMetrics(
+    const YAML::Node& document_node) {
+  if (!document_node["model_metrics"]) {
+    return;
+  }
+
+  const YAML::Node model_metrics_node = document_node["model_metrics"];
+
+  if (model_metrics_node["counter"]) {
+    const std::vector<MetricConfiguration> counter_metrics =
+        ParseMetrics(model_metrics_node["counter"], MetricType::COUNTER);
+    model_metrics.insert(model_metrics.end(), counter_metrics.begin(),
+                         counter_metrics.end());
+  }
+
+  if (model_metrics_node["gauge"]) {
+    const std::vector<MetricConfiguration> gauge_metrics =
+        ParseMetrics(model_metrics_node["gauge"], MetricType::GAUGE);
+    model_metrics.insert(model_metrics.end(), gauge_metrics.begin(),
+                         gauge_metrics.end());
+  }
+
+  if (model_metrics_node["histogram"]) {
+    const std::vector<MetricConfiguration> histogram_metrics =
+        ParseMetrics(model_metrics_node["histogram"], MetricType::HISTOGRAM);
+    model_metrics.insert(model_metrics.end(), histogram_metrics.begin(),
+                         histogram_metrics.end());
+  }
+}
+
+std::vector<MetricConfiguration> YAMLMetricsConfigurationHandler::ParseMetrics(
+    const YAML::Node& metrics_list_node, const MetricType& metric_type) {
+  std::vector<MetricConfiguration> metrics_config =
+      metrics_list_node.as<std::vector<MetricConfiguration>>();
+  for (auto& metric : metrics_config) {
+    std::set<std::string> metric_dimensions_set{};
+    for (const auto& name : metric.dimension_names) {
+      if (dimension_names.find(name) == dimension_names.end()) {
+        std::string error_message =
+            "Dimension \"" + name +
+            "\" associated with metric: " + metric.name +
+            " not defined under central \"dimensions\" key in configuration";
+        throw YAML::Exception(YAML::Mark::null_mark(), error_message);
+      }
+      if (metric_dimensions_set.insert(name).second == false) {
+        std::string error_message =
+            "Dimensions for metric: " + metric.name + " must be unique";
+        throw YAML::Exception(YAML::Mark::null_mark(), error_message);
+      }
+    }
+
+    metric.type = metric_type;
+  }
+
+  return metrics_config;
+}
+}  // namespace torchserve
+
+namespace YAML {
```

Comment:
What is this namespace used for?

Reply:
The decode method defined below is used to conveniently parse the list of metrics specifications under each metric type `counter`, `gauge` and `histogram` in the configuration.
Reference: https://github.com/jbeder/yaml-cpp/wiki/Tutorial#converting-tofrom-native-data-types
Usage: `std::vector<MetricConfiguration> metrics_config = metrics_list_node.as<std::vector<MetricConfiguration>>();`